### DISCUSSION TITLE
How to remove the history of a document and have single snapshot file?

### DISCUSSION THREAD

**Discussion by alienself** [automerge/automerge#1520] [General]
Hi,

I would like to know how to remove the history of a document, after a certain amount of edit, the load time become noticeable so I would like to remove the history to compact all the incremental files into a single snapshot file.
How can I do that?

I tried something like below by creating new handle and commiting changes back in however it creates an incremental file, which is not what I want.

```ts
// create new document
const newHandle = serverRepo.create();

// apply changes to new document's handle
newHandle.change((d: any) => {
    for (const key in oldDocument) {
        const value = oldDocument[key];
        d[key] = value;
    }
});

// remove old document
serverRepo.delete(oldHandle.url);
```

**Comment by pvh** [comment]
I usually do something like
```
const newHandle = serverRepo.create();
newHandle.change( d => {
   Object.assign(d, JSON.parse(JSON.stringify(oldDocument)))
}|)
```

This is from memory, but roughly -- Object.assign splats out the keys into the root object and bouncing through JSON strips off any metadata.

**Comment by pvh** [comment]
I'll leave this open for addition to the docs cookbook or something.

**Comment by johnspurlock** [comment]
Is there any way to compact an existing document _without_ creating a new one?

**Comment by pvh** [comment]
"Compact" could mean a few different things here -- one is whether you can reduce the on-disk size of an Automerge document. This happens automatically when you call `save()`. It quite literally compacts the history into a very efficient columnar representation.

Discarding the history is not currently possible without replacing the document: if you don't have the same history, you can't merge the same changes. That's because we use a git-like commit hash for each change to ensure integrity. Otherwise, you can get into gnarly situations where multiple nodes believe diverging things about your data yet keep trying to exchange deltas. 

In `git`, you can kind of pretend the problem away with a rebase/force push. This can cause problems if anyone else has already downloaded those changes but if not, you can usually get away with it. Unfortunately, Automerge changes are generally much more granular and tend towards real-time synchronization. In particular, a well-behaved Automerge application is likely synchronizing data in real-time with another process all the time, whether that's a local backup node, a team's synchronization server, or just another process in the same browser.

We've discussed a few longer-term directions here. 

One is akin to a "shallow clone", which is to allow a peer to fetch just a particular snapshot without the dependencies. They'll have to trust the peer that sent it not to lie to them (since without the history, they can't prove to themselves that the hashes are correct) but could continue to receive streaming changes and in the event of receiving a "back-dated" / pre-snapshot change could request additional context from that sending peer.  This would mostly be useful for optimizing initial document load speed or creating lightweight clients that don't need to remember their past.

The other is what we're calling a "cherry pick", which is to say that we want to be able to squash a bunch of changes into a commit in a minimal way *as* a commit itself. This could allow, for example, sending a patch to an upstream project based on an interactive pair-programming session but without all the meandering details along the way.

Both of these are obviously useful, related, and not yet implemented, but give you the flavor of the directions we're considering. Input on what the problems you're trying to solve are and what kinds of constraints you're operating under are always helpful, and of course, if this is something your business would like to sponsor it helps us pull things forward when we have someone we know wants a feature enough to contribute to its development.

Hope that helps!

**Comment by johnspurlock** [comment]
Very helpful, thanks - I'm still in the evaluation phase, wondering if automerge would be a good fit for a new project.  In my case, the peers are only occasionally all online at the same time, and may have a large number of small changes while disconnected - so the detailed history of the changes while offline are not as important as the "squashed" version.  But while they are connected, I would want a realtime level of synchronization, with small deltas/syncs etc

Wondering if I can maybe just defined the "document" as the changes during only the connected period, and just a more traditional store when disconnected, still would be nice to send a single "snapshot" for these cases, and also prune the unnecessary history.

**Comment by Iceburgino** [comment]
@pvh Shallow clone is a really nice feature. In our case, users can work on a single document for years, that's why we'd like to have constant load times. It's something similar to mercurial, where it periodically snapshots the entire document, to avoid replaying the entire history on checkout.

**Comment by pvh** [comment]
Yes, it's somewhere on the wishlist for us too. I don't know *when* it will happen, but it's on the list. This is likely a feature that a sponsor could help us pull forward with extra funding.
